Allow Sandbox to function as an open web application as well as a SMART app to enable developers to use authorized fhir servers, launch with context, and send access tokens to CDS services.

zplata opened the pull request:
Issue: https://github.com/cds-hooks/sandbox/issues/48

Currently, the Sandbox is an open stand-alone tool that developers can use by just hitting the sandbox.cds-hooks.org link. This method uses an open smarthealthit FHIR server endpoint to deal with FHIR transactions (as well as launching SMART apps). We need a way to allow the Sandbox to run against secured FHIR servers to align with the 1.0 spec security model.

To remedy part of this problem, we could allow the Sandbox to handle the embedded SMART launch flow, thus making the Sandbox a possible SMART app. This means that the CDS Hooks Sandbox can be launched from a developer sandbox by HSPC or any other EHR sandbox environment, where the CDS Hooks Sandbox gets passed launch context, and can query the respective secured FHIR server. Additionally, the Sandbox should still be able to handle being launched as an "open" web application.

Currently, with this change, here's what is being read into the Sandbox:
- Patient ID
- User ID
- FHIR Server (and access token)

Notes:
- If the CDS Hooks Sandbox is launched as a SMART app, the option to "change" the FHIR server will not be available. 
- The "open" web application version does not yet support using secured FHIR servers, as the main intent of this change is to convert the Sandbox into a SMART app
- The client ID used for now is hooked up to my own personal HSPC sandbox instance, but will be switched to whatever client ID is registered in the future CDS Hooks HSPC sandbox

zplata commented:
To test this change, accept my invite to share my personal HSPC sandbox instance (your email). Under launch scenarios, launch the scenario called "Use me to test the Sandbox as a SMART app".

zplata commented:
Added enhancements to allow this Sandbox to go through the launch workflow against an open FHIR endpoint.

koushic88 commented on src/scripts/components/App.js:
```diff
@@ -25,10 +25,17 @@ function bodyClick(e) {
-  type: ActionTypes.NEW_HASH_STATE,
-  hash: window.location.hash.slice(1) ? JSON.parse(window.location.hash.slice(1)) : {}
-})
+if(window.location.hash.slice(1).indexOf('state=') < 0) {
```
can we do nil check for window.location.hash.slice(1) and then call indexOf() ?

koushic88 commented on src/scripts/components/App.js:
```diff
@@ -25,10 +25,17 @@ function bodyClick(e) {
+if(window.location.hash.slice(1).indexOf('state=') < 0) {
+  AppDispatcher.dispatch({
+    type: ActionTypes.NEW_HASH_STATE,
+    hash: window.location.hash.slice(1) ? JSON.parse(window.location.hash.slice(1)) : {}
```
If the window.location.hash.slice(1) is nil or undefined, it will go to the else condition. Do we need this check here?

zplata replied:
@koushic88 `window.location.hash.slice(1)` will always return an empty string if it doesn't match some value in the URL, and will actually end up in this block of code. This is checking if the value of `window.location.hash.slice(1)` is an actual string with value, and if so, parsing it accordingly.

kpshek commented on src/scripts/main.js:
```diff
@@ -9,5 +9,28 @@ require('!style-loader!css-loader!codemirror/addon/lint/lint.css');
+  () => {
+    var qs = querystring.parse(window.location.search.slice(1));
+    var fhirContext = {
+      patient: qs.patientId || "SMART-1288992",
```
It seems like this is one of a few places where we reference the default patient id (SMART-1288992). Can we break this out into a global that we reference, allowing us to set the default in just one place?

kpshek replied:
fyi, this is my last comment -- everything else looks good

zplata replied:
The app uses the Flux library, which allows for multiple stores to keep state. There's a `FhirServerStore` which stores the patient (I set the patient in context below this code for that `FhirServerStore`). I stripped this logic out and only updated the patient in context if the Sandbox is going through a secured smart launch. 

See update on files changed

kpshek commented on src/scripts/stores/DecisionStore.js:
```diff
@@ -77,6 +78,9 @@ function _hooksChanged() {
         url: context.baseUrl + '/' + fillTemplate(url, context),
+        headers: (CDS_SMART_OBJ.accessToken && CDS_SMART_OBJ.accessToken.access_token) ?
+          {'Authorization': 'Bearer ' + CDS_SMART_OBJ.accessToken.access_token,
+           'Accept': 'application/json+fhir'} : {'Accept': 'application/json+fhir'},
```
I'm missing something -- what's the 2nd Accept key/value here?

zplata replied:
I'll break this logic out in a variable once I clean this up. But this is basically setting the authorization and accept header if a access token is present, otherwise, just sets the accept header.

kpshek replied:
Ah...I missed that ternary operator. 😄

kpshek commented on build/launch.html:
```diff
@@ -0,0 +1,19 @@
+  </head>
+  <body>
+    <script>
+      // TODO: Replace this client_id with registered app client_id from specified sandbox
```
Do we need this TODO here?

zplata replied:
We can take it out. I just put it here as a todo to replace this current client ID (from my personal sandbox) with the one that's registered on the new HSPC Sandbox.

zplata commented:
Addressed comments, but a few things: 

1) Should we display the oauth property in the Service request (since we display that now) that contains the access token and scopes? Would that be useful for those developing?
2) I've updated the client ID in the launch.html for the app registered on my shared HSPC Sandbox instance. This app will point to engineering.com/cds-hooks-sandbox/launch.html. When we PR this to the host cds-hooks/sandbox repo, I will change the client ID to the app registered in the new HSPC CDS Hooks Sandbox instance. 

@kpshek

kpshek commented:
> Should we display the oauth property in the Service request (since we display that now) that contains the access token and scopes? Would that be useful for those developing?

Yes, we should.

> I've updated the client ID in the launch.html for the app registered on my shared HSPC Sandbox instance. This app will point to engineering.com/cds-hooks-sandbox/launch.html. When we PR this to the host cds-hooks/sandbox repo, I will change the client ID to the app registered in the new HSPC CDS Hooks Sandbox instance.

Cool. As an aside, if developers want to configure their own secure FHIR server (eg from HSPC or another vendor), then they will also need to change the OAuth 2 client id.
